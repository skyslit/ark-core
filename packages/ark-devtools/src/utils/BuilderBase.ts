import webpack, {Configuration, Stats} from 'webpack';
import {EventEmitter} from 'events';

type Mode = 'development' | 'production';
export type ConfigurationOptions = {
  mode: Mode,
  cwd: string,
}
/**
 * Wrapper for Webpack
 */
export class BuilderBase extends EventEmitter {
  private compiler: webpack.Compiler;
  /**
   * Creates a new builder base instance
   * @param {EventEmitterOptions} options
   */
  constructor(options?: any) {
    super(options);
  }

  /**
   * Start build process
   * @param {ConfigurationOptions} opts
   * @param {any=} ifs Input filesystem
   * @param {any=} ofs Output filesystem
   * @param {any=} wfs Watch filesystem
   */
  build(opts: ConfigurationOptions, ifs?: any, ofs?: any, wfs?: any) {
    const buildConfiguration = this.getConfiguration(
        Object.assign<
          ConfigurationOptions,
          Partial<ConfigurationOptions>
        >({
          mode: 'production',
          cwd: null,
        }, opts)
    );
    if (!buildConfiguration) {
      throw new Error(
          'webpack configuration should not be null'
      );
    }
    this.compiler = webpack(buildConfiguration);
    if (ifs) {
      this.compiler.inputFileSystem = ifs;
    }
    if (ofs) {
      this.compiler.outputFileSystem = ofs;
    }
    if (wfs) {
      this.compiler.watchFileSystem = wfs;
    }
    if (opts.mode === 'development') {
      this.compiler.watch({}, this.handler.bind(this));
    } else if (opts.mode === 'production') {
      this.compiler.run(this.handler.bind(this));
    }
  }

  /**
   * Supported events
   * @return {string[]}
   */
  eventNames(): string[] {
    return [
      'success',
      'warning',
      'error',
    ];
  }

  /**
   * Gets configuration
   * @param {ConfigurationOptions} opts
   * @return {Configuration}
   */
  getConfiguration(opts: ConfigurationOptions): Configuration {
    return null;
  }

  /**
   * Handler
   * @param {Error} err
   * @param {Stats} result
   */
  private handler(err?: Error, result?: Stats): void {
    if (err) {
      this.emit('error', [{
        message: err.message,
      }]);
    } else {
      if (result.hasErrors()) {
        this.emit('error',
            result.compilation.errors,
            result.compilation,
            result
        );
      } else if (result.hasWarnings()) {
        this.emit('warning',
            result.compilation.warnings,
            result.compilation,
            result
        );
      } else {
        this.emit('success', result.compilation, result);
      }
    }
  }
}
