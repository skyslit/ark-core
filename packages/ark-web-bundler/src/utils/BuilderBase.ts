import webpack, {Compilation, Configuration, Stats} from 'webpack';
import {EventEmitter} from 'events';

type Mode = 'development' | 'production';
export type ConfigurationOptions = {
  mode: Mode,
  cwd: string,
}
/**
 * Wrapper for Webpack
 */
export class BuilderBase extends EventEmitter {
  private compiler: webpack.Compiler;
  /**
   * Creates a new builder base instance
   * @param {EventEmitterOptions} options
   */
  constructor(options?: any) {
    super(options);
  }

  /**
   * Start build process
   * @param {ConfigurationOptions} opts
   * @param {any=} fs
   */
  build(opts: ConfigurationOptions, fs?: any) {
    const buildConfiguration = this.getConfiguration(
        Object.assign<
          ConfigurationOptions,
          Partial<ConfigurationOptions>
        >({
          mode: 'production',
          cwd: null,
        }, opts)
    );
    console.log(buildConfiguration);
    if (!buildConfiguration) {
      throw new Error(
          'webpack configuration should not be null'
      );
    }
    this.compiler = webpack(buildConfiguration);
    // if (fs) {
    //   this.compiler.inputFileSystem = fs;
    //   this.compiler.outputFileSystem = fs;
    //   this.compiler.watchFileSystem = fs;
    // }
    if (opts.mode === 'development') {
      this.compiler.watch({}, this.watchHandler.bind(this));
    } else if (opts.mode === 'production') {
      this.compiler.compile(this.handler.bind(this));
    }
  }

  /**
   * Supported events
   * @return {string[]}
   */
  eventNames(): string[] {
    return [
      'success',
      'warning',
      'error',
    ];
  }

  /**
   * Gets configuration
   * @param {ConfigurationOptions} opts
   * @return {Configuration}
   */
  getConfiguration(opts: ConfigurationOptions): Configuration {
    return null;
  }

  /**
   * Handler
   * @param {Error} err
   * @param {Stats} result
   */
  private watchHandler(err?: Error, result?: Stats): void {
    if (err) {
      this.emit('error', [{
        message: err.message,
      }]);
    } else {
      if (result.hasErrors()) {
        this.emit('error',
            result.compilation.errors,
            result.compilation,
            result
        );
      } else if (result.hasWarnings()) {
        this.emit('warning',
            result.compilation.warnings,
            result.compilation,
            result
        );
      } else {
        this.emit('success', result.compilation, result);
      }
    }
  }

  /**
   * Handler
   * @param {Error} err
   * @param {Stats} result
   */
  private handler(err?: Error, result?: Compilation): void {
    if (err) {
      this.emit('error', [{
        message: err.message,
      }]);
    } else {
      if (result.errors.length > 0) {
        this.emit('error',
            result.errors,
            result
        );
      } else if (result.warnings.length > 0) {
        this.emit('warning',
            result.warnings,
            result
        );
      } else {
        this.emit('success', result);
      }
    }
  }
}
