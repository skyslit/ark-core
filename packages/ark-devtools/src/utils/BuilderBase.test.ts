/* eslint-disable require-jsdoc */
import {Compilation, Configuration} from 'webpack';
import {BuilderBase, ConfigurationOptions} from './BuilderBase';
import path from 'path';
import memfs from 'memfs';
import {Union} from 'unionfs';
import * as fs from 'fs';

describe('builder base configuration', () => {
  class SampleBuilder extends BuilderBase {
    getConfiguration({cwd, mode}: ConfigurationOptions): Configuration {
      return {
        mode,
      };
    }
  }

  test('getFullyQualifiedConfiguration() fn', () => {
    const builderInstance = new SampleBuilder();
    const configuration = builderInstance.getConfiguration({
      cwd: '/test',
      mode: 'production',
    });
    expect(configuration.mode).toEqual('production');
  });
});

describe('build stage: production', () => {
  class SampleBuilder extends BuilderBase {
    private entryPathname: string;
    constructor(entryPath: string) {
      super();
      this.entryPathname = entryPath;
    }

    getConfiguration({cwd, mode}: ConfigurationOptions): Configuration {
      return {
        mode,
        entry: this.entryPathname,
        output: {
          filename: 'main.js',
          path: path.resolve(cwd, 'build'),
        },
        module: {
          rules: [
            {
              test: /\.(ts|tsx|js|jsx)$/,
              use: [
                {
                  loader: path.resolve(
                      __dirname, '../../node_modules', 'babel-loader'
                  ),
                },
              ],
            },
          ],
        },
      };
    }
  }

  const cwd: string = process.cwd();
  const testRoot: string = '__webpack__test';
  let vol: any;
  let outputFileSystem: any;
  let inputFileSystem: any;

  beforeEach(() => {
    // Setup Output Filesystem
    vol = memfs.Volume.fromJSON({
      [`${testRoot}/main.server.ts`]: `console.log('Server program');`,
      [`${testRoot}/main-error.server.ts`]: `console.log('Server program);`,
      [`${testRoot}/dashboard.client.js`]:
        `console.log('Dashboard client program');`,
      [`${testRoot}/admin.client.js`]: `console.log('Admin client program');`,
    }, cwd);
    outputFileSystem = memfs.createFsFromVolume(vol);

    // Setup Input Filesystem
    inputFileSystem = new Union();
    inputFileSystem.use(fs).use(vol as any);
  });

  test('success operation', (done) => {
    const builderInstance = new SampleBuilder(
        path.join(cwd, testRoot, 'main.server.ts')
    );
    builderInstance.on('success', (compilation: Compilation) => {
      try {
        const buildOutput: string = outputFileSystem.readFileSync(
            path.join(cwd, 'build', 'main.js'),
            'utf-8'
        );
        expect(buildOutput).toContain('Server program');
        expect(buildOutput).toMatchSnapshot();
        done();
      } catch (e) {
        done(e);
      }
    });
    builderInstance.on('warning', (warnings) => {
      done(new Error('Warnings should not be thrown'));
    });
    builderInstance.on('error', (errors) => {
      done(new Error('Error should not be thrown'));
    });
    builderInstance.build({
      mode: 'production',
      cwd: process.cwd(),
    }, inputFileSystem, outputFileSystem);
  });
  // test('warnings operation', () => {

  // });
  test('error operation', (done) => {
    const builderInstance = new SampleBuilder(
        path.join(cwd, testRoot, 'main-error.server.ts')
    );
    builderInstance.on('success', () => {
      done(new Error('Should not trigger success'));
    });
    builderInstance.on('warning', (warnings) => {
      done(new Error('Should not trigger warning'));
    });
    builderInstance.on('error', (errors) => {
      expect(errors).toHaveLength(1);
      done();
    });
    builderInstance.build({
      mode: 'production',
      cwd: process.cwd(),
    }, inputFileSystem, outputFileSystem);
  });
});
