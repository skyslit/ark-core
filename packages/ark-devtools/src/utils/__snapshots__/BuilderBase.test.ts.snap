// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build stage: production success operation 1`] = `"console.log(\\"Server program\\");"`;
